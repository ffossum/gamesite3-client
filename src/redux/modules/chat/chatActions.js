/* @flow */
export const JOIN_CHANNEL = 'chat/join channel';
export const LEAVE_CHANNEL = 'chat/leave channel';
export const SEND_MESSAGE = 'chat/send message';
export const SEND_GAME_MESSAGE = 'chat/send game message';
export const RECEIVE_MESSAGE = 'chat/receive message';
export const CLEAR_CHAT = 'chat/clear';

export type ChannelName = string;
export type UserMessage = {
  ch: ChannelName,
  txt: string,
  uid: string,
};

export type JoinChannelAction = {
  type: typeof JOIN_CHANNEL,
  payload: string,
};
export function joinChannel(channelName: ChannelName): JoinChannelAction {
  return {
    type: JOIN_CHANNEL,
    payload: channelName,
  };
}

export type LeaveChannelAction = {
  type: typeof LEAVE_CHANNEL,
  payload: string,
};
export function leaveChannel(channelName: ChannelName): LeaveChannelAction {
  return {
    type: LEAVE_CHANNEL,
    payload: channelName,
  };
}

export type SendMessageAction = {
  type: typeof SEND_MESSAGE,
  payload: UserMessage,
};
export function sendMessage(
  userId: string,
  channelName: string,
  messageText: string,
): SendMessageAction {
  return {
    type: SEND_MESSAGE,
    payload: {
      ch: channelName,
      txt: messageText,
      uid: userId,
    },
  };
}

export type SendGameMessageAction = {
  type: typeof SEND_GAME_MESSAGE,
  payload: {
    players: string[],
    gameId: string,
    userId: string,
    text: string,
  },
};
export function sendGameMessage(
  userId: string,
  gameId: string,
  players: string[],
  text: string,
) {
  return {
    type: SEND_GAME_MESSAGE,
    payload: {
      userId,
      gameId,
      players,
      text,
    },
  };
}
export type ReceiveMessageAction = {
  type: typeof RECEIVE_MESSAGE,
  payload: {
    msg: UserMessage,
    time: string,
  },
};
export function receiveMessage(
  msg: UserMessage,
  time: string,
): ReceiveMessageAction {
  return {
    type: RECEIVE_MESSAGE,
    payload: {
      msg,
      time,
    },
  };
}

export type ClearChatAction = {
  type: typeof CLEAR_CHAT,
  payload: string,
};
export function clearChat(channelName: string) {
  return {
    type: CLEAR_CHAT,
    payload: channelName,
  };
}

export type ChatAction =
  | JoinChannelAction
  | LeaveChannelAction
  | SendMessageAction
  | SendGameMessageAction
  | ReceiveMessageAction
  | ClearChatAction;
