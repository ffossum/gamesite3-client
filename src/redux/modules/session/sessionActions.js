/* @flow */
export const AUTHENTICATED_USER = 'user/authenticated';

export type AuthenticatedUserAction = {
  type: 'user/authenticated',
  payload: PublicUserData,
};
export function authenticatedUser(user: PublicUserData) {
  return {
    type: AUTHENTICATED_USER,
    payload: user,
  };
}
