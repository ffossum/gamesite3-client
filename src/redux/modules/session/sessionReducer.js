/* @flow */
import type { Action } from '../../actions';
import { REGISTRATION_SUCCESS } from '../registration/registrationActions';

export type SessionUser = {
  id: string,
  username: string,
};
export type SessionState = {
  user: ?SessionUser,
};
const initialState: SessionState = {
  user: null,
};

export default function sessionReducer(
  state: SessionState = initialState,
  action: Action
) {
  switch (action.type) {
    case REGISTRATION_SUCCESS:
      return {
        ...state,
        user: action.payload,
      };
    default:
      return state;
  }
}
