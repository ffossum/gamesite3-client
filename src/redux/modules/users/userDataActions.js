/* @flow */
export const FETCHED_USER_DATA = 'data/fetched user(s)';

type FetchedUserDataAction = {
  type: typeof FETCHED_USER_DATA,
  payload: PublicUserData[],
};
export function fetchedUserData(
  users: PublicUserData[],
): FetchedUserDataAction {
  return {
    type: FETCHED_USER_DATA,
    payload: users,
  };
}

export type UserDataAction = FetchedUserDataAction;
