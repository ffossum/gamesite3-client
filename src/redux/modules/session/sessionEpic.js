/* @flow */
import { Observable } from 'rxjs';
import { AUTHENTICATED_USER } from './sessionActions';
import type { AuthenticatedUserAction } from './sessionActions';
import type { Store } from 'redux';
import type DeepstreamClient from '../../deepstreamClient';
import { playerJoined, playerLeft } from '../games/gameRoomActions';

type Dependencies = {
  deepstreamClient: DeepstreamClient,
};

export default function sessionEpic(
  action$: Observable<*>,
  store: Store<*, *>,
  { deepstreamClient }: Dependencies,
) {
  return action$
    .filter(action => action.type === AUTHENTICATED_USER)
    .switchMap((action: AuthenticatedUserAction) => {
      return deepstreamClient
        .subscribe('user:' + action.payload.id)
        .flatMap(data => {
          switch (data.t) {
            case 'player-joined': {
              const gameId = data.p.gid;
              const userId = data.p.uid;
              return [playerJoined(userId, gameId)];
            }
            case 'player-left': {
              const gameId = data.p.gid;
              const userId = data.p.uid;
              return [playerLeft(userId, gameId)];
            }
            default:
              return [];
          }
        });
    });
}
