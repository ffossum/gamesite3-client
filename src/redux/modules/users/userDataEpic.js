/* @flow */
import { stringify } from 'query-string';
import { fetchedUserData } from './userDataActions';
import { RECEIVE_MESSAGE } from '../chat/chatActions';
import type { ReceiveMessageAction } from '../chat/chatActions';

type Dependencies = {
  ajax: any,
};
export default function userDataEpic(
  action$: ActionsObservable<*>,
  store: Store<*>,
  { ajax }: Dependencies
) {
  return action$
    .ofType(RECEIVE_MESSAGE)
    .map((action: ReceiveMessageAction) => action.payload.msg.uid)
    .distinct()
    .bufferTime(100)
    .filter(userIds => userIds.length)
    .flatMap(userId => {
      const queryString = stringify({
        id: [userId],
      });
      return ajax.getJSON(`/api/users?${queryString}`);
    })
    .map(users => fetchedUserData(users));
}
