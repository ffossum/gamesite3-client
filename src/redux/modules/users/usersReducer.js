/* @flow */
import type { Action } from '../../actions';
import { FETCHED_USER_DATA } from './userDataActions';
import { LOGIN_SUCCESS } from '../login/loginActions';

export type UsersState = {
  [string]: PublicUserData,
};

const initialState = {};

export default function usersReducer(
  state: UsersState = initialState,
  action: Action,
) {
  switch (action.type) {
    case FETCHED_USER_DATA: {
      const newUsers = {};
      action.payload.forEach(user => {
        newUsers[user.id] = user;
      });

      return {
        ...state,
        ...newUsers,
      };
    }
    case LOGIN_SUCCESS: {
      const user = action.payload;
      return {
        ...state,
        [user.id]: user,
      };
    }
    default:
      return state;
  }
}
