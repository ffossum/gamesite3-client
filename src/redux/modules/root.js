/* @flow */
import { combineReducers } from 'redux';
import { combineEpics } from 'redux-observable';
import registration from './registration/registrationReducer';
import type { RegistrationState } from './registration/registrationReducer';
import registrationEpic from './registration/registrationEpic';
import session from './session/sessionReducer';
import type { SessionState } from './session/sessionReducer';
import sessionEpic from './session/sessionEpic';

import login from './login/loginReducer';
import type { LoginState } from './login/loginReducer';
import loginEpic from './login/loginEpic';

import chat from './chat/chatReducer';
import type { ChatState } from './chat/chatReducer';
import chatEpic from './chat/chatEpic';

import users from './users/usersReducer';
import userDataEpic from './users/userDataEpic';
import type { UsersState } from './users/usersReducer';

import lobby from './lobby/lobbyReducer';
import lobbyEpic from './lobby/lobbyEpic';
import type { LobbyState } from './lobby/lobbyReducer';

import games from './games/gamesReducer';
import type { GamesState } from './games/gamesReducer';
import gameRoomEpic from './games/gameRoomEpic';

export type State = {
  chat: ChatState,
  login: LoginState,
  registration: RegistrationState,
  session: SessionState,
  users: UsersState,
  lobby: LobbyState,
  games: GamesState,
};

export const rootReducer = combineReducers({
  chat,
  login,
  registration,
  session,
  users,
  lobby,
  games,
});

export const rootEpic = combineEpics(
  chatEpic,
  loginEpic,
  registrationEpic,
  sessionEpic,
  userDataEpic,
  lobbyEpic,
  gameRoomEpic,
);
