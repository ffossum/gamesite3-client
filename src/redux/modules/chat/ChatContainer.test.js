/* @flow */
/* eslint-env jest */
import { createStore } from 'redux';
import { authenticatedUser } from '../session/sessionActions';
import { rootReducer } from '../root';
import { sendMessage, receiveMessage } from './chatActions';
import { fetchedUserData } from '../users/userDataActions';
import {
  mapStateToProps as createMapStateToProps,
  mapDispatchToProps,
  mergeProps,
} from './ChatContainer';

describe('chat container', () => {
  let store;
  let dispatch;
  let mapStateToProps;
  const ownProps = { channelName: 'channel' };

  beforeEach(() => {
    store = createStore(rootReducer);
    dispatch = jest.fn();
    mapStateToProps = createMapStateToProps(undefined, ownProps);
  });

  test('does not dispatch action when logged out user tries to send message', () => {
    const stateProps = mapStateToProps(store.getState(), ownProps);
    const dispatchProps = mapDispatchToProps(dispatch);

    const props = mergeProps(stateProps, dispatchProps, ownProps);

    props.sendMessage('hello');

    expect(dispatch).not.toHaveBeenCalled();
  });

  test('dispatches action when logged in user sends message', () => {
    store.dispatch(
      authenticatedUser({
        email: 'asdf@asdf.com',
        id: 'asdf-id',
        username: 'asdf',
      }),
    );

    const stateProps = mapStateToProps(store.getState(), ownProps);
    const dispatchProps = mapDispatchToProps(dispatch);

    const props = mergeProps(stateProps, dispatchProps, ownProps);

    props.sendMessage('hello');

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(
      sendMessage('asdf-id', 'channel', 'hello'),
    );
  });

  test('creates denormalized message array', () => {
    store.dispatch(
      receiveMessage(
        { ch: 'channel', txt: 'hey', uid: 'zxcv-id' },
        '2017-06-08T02:07:37.605Z',
      ),
    );
    store.dispatch(
      receiveMessage(
        { ch: 'channel', txt: 'hi', uid: 'qwer-id' },
        '2017-06-08T02:07:37.605Z',
      ),
    );

    store.dispatch(
      fetchedUserData([
        { id: 'qwer-id', username: 'qwer' },
        { id: 'zxcv-id', username: 'zxcv' },
      ]),
    );

    const stateProps = mapStateToProps(store.getState(), ownProps);

    expect(stateProps).toMatchSnapshot();
  });
});
