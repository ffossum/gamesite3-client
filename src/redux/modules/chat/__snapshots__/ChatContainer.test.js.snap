// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat container creates denormalized message array 1`] = `
Object {
  "messages": Array [
    Object {
      "text": "hey",
      "time": "2017-06-08T02:07:37.605Z",
      "user": Object {
        "id": "zxcv-id",
        "username": "zxcv",
      },
    },
    Object {
      "text": "hi",
      "time": "2017-06-08T02:07:37.605Z",
      "user": Object {
        "id": "qwer-id",
        "username": "qwer",
      },
    },
  ],
  "user": null,
}
`;
